### DISCUSSION TITLE
Cutoff dimension for sparse matrix operation eggs

### DISCUSSION THREAD

**Discussion by tony-teng-yt377** [Part-II-Computational-Physics/.github#35] [General]
For optional task 10, I need to use the ground state vector obtained from eigsh(SparseHamiltonian), but the latter requires me to set a cutoff dimension k < N, so I have an unmatched dimension between my M_x operator and dimension-reduced ground state eigenvector. (So I can't quite act magnetisation operator on the ground state eigenvector) How should I resolve this issue?

Many thanks! (sorry, I meant to write operation eigsh)

**Comment by AprilSweettooth** [comment]
Same problem here and I tried to get the ground state from complete hamiltonian and computed the expectation with sparse matrix. Not sure if it's applicable ...

**Reply by tony-teng-yt377** [reply]
Yeah that's exactly what I am trying, though my magnitization does not look like something I would expect..

**Comment by JNHallen** [comment]
Hi! The optional input k in scipy.sparse.linalg.eigsh determines how many eigenvalues and eigenvectors you calculate. You cannot calculate all $N$ eigenvectors, but you can find a couple of them. For this Hamiltonian the groundstate is at most double degenerate, so you should never need to calculate more than two eigenvectors to find the groundstate. Does that clear things up?

**Reply by tony-teng-yt377** [reply]
Hi thanks for reply, the problem is that k not only determines how many eigenvectors you calculate, but it also somehow determines the dimension of the eigenvector. So if I do k = 6 on a 13 * 13 sparse matrix, the output vectors will be only of dimension 6 but not 13

**Reply by JNHallen** [reply]
If you do:

E_0, psi = spsl.eigsh(SparseHamiltonian(N, h=h), k=2, which='SA')

that should give you psi as an array with shape ($2^N$, $2$). This corresponds to the two lowest energy eigenvectors, each with $2^N$ elements. Does that not work?

**Reply by tony-teng-yt377** [reply]
Oh wow, yeah that works, which = "SA" makes all the difference! Thank you very much1

**Reply by JNHallen** [reply]
Yes, sorry I realise now that I should have specifically mentioned that which="SA" forces eigsh to look for the smallest amplitude eigenvalues, which are the ones we want.

**Reply by tony-teng-yt377** [reply]
I think the main issue without which = "SA" is not even the order of the eigenvalue, but somehow it cuts off a few components of each eigenvectors. But thanks anyway

**Reply by JNHallen** [reply]
Are you sure that is the case? When I for example run:

E_0, psi = spsl.eigsh(SparseHamiltonian(8, h=1), k=k) #, which='SA')

I do get psi as an output array of shape ($2^8$, k), which is the correct dimensions. (Without "which='SA'" there are some convergence issues for larger N though.)

**Reply by tony-teng-yt377** [reply]
Oh sorry I was wrong, after I tried it again, it seems to be the right dimension, it's coz I have to transpose it to get the correct row vectors. Thanks!
